### PR TITLE
[Upgrade] Upgrade Sentinel and Enhance

### LINKED ISSUE #1168: [Sentinel] version update & use SentinelWebInterceptor instead of CommonFilter
**Which Component**

Sentinel

**Is your feature request related to a problem? Please describe.**

1. Version upgrade to 1.7.1
2. Use [SentinelWebInterceptor](https://github.com/alibaba/Sentinel/blob/master/sentinel-adapter/sentinel-spring-webmvc-adapter/src/main/java/com/alibaba/csp/sentinel/adapter/spring/webmvc/SentinelWebInterceptor.java) instead of [CommonFilter](https://github.com/alibaba/Sentinel/blob/master/sentinel-adapter/sentinel-web-servlet/src/main/java/com/alibaba/csp/sentinel/adapter/servlet/CommonFilter.java)

### PR DESCRIPTION
**Pull request by yuhuangbin**
### Describe what this PR does / why we need it
- Upgrade Sentinel to 1.7.1
- Introduce `SentinelWebInterceptor` to replace `CommonFilter`

### Does this pull request fix one issue?
fixed #1168

### CODE REVIEW

**Comment by fangjian0423** [comment]
This is the comment is HandlerInterceptor: 

> HandlerInterceptor is basically similar to a Servlet Filter, but in contrast to the latter it just allows custom pre-processing with the option of prohibiting the execution of the handler itself, and custom post-processing.

I think keep the filed name `filter` of SentinelProperties could be accepted.

**Comment by sczyh30** [comment]
> This is the comment is HandlerInterceptor:
> 
> > HandlerInterceptor is basically similar to a Servlet Filter, but in contrast to the latter it just allows custom pre-processing with the option of prohibiting the execution of the handler itself, and custom post-processing.
> 
> I think keep the filed name `filter` of SentinelProperties could be accepted.

+1, and this could also keep compatibility.

By the way, any ideas for detecting whether the user has already introduced the legacy CommonFilter or the interceptor itself (to avoid duplication)?

**Comment by fangjian0423** [comment]
> > This is the comment is HandlerInterceptor:
> > > HandlerInterceptor is basically similar to a Servlet Filter, but in contrast to the latter it just allows custom pre-processing with the option of prohibiting the execution of the handler itself, and custom post-processing.
> > 
> > 
> > I think keep the filed name `filter` of SentinelProperties could be accepted.
> 
> +1, and this could also keep compatibility.
> 
> By the way, any ideas for detecting whether the user has already introduced the legacy CommonFilter or the interceptor itself (to avoid duplication)?

1. comments in `SentinelProperties` describe how we use filter.

image

So all the fileds comment in SentinelProperties$Filter should by modified.

2. The logs printed changed from "[Sentinel Starter] register Sentinel CommonFilter with urlPatterns: ..." to "[Sentinel Starter] register SentinelWebInterceptor with urlPatterns: ..."

Any better ideas?

**Comment by fangjian0423** [comment]
> The `spring.cloud.sentinel.filter.enabled` lost effect.
> And shoud remove `sentinel-web-servlet` dependency? @fangjian0423

1. yes, `spring.cloud.sentinel.filter.enabled` should be remain and do its effect as before
2. I think it could be removed @sczyh30

**Comment by yuhuangbin** [comment]
> 2\. hink it could be remove

If `sentinel-web-servlet` dependency is removed, `spring.cloud.sentinel.servlet` still needed?

**Comment by sczyh30** [comment]
> > The `spring.cloud.sentinel.filter.enabled` lost effect.
> > And shoud remove `sentinel-web-servlet` dependency? @fangjian0423
> 
> 1. yes, `spring.cloud.sentinel.filter.enabled` should be remain and do its effect as before
> 2. I think it could be removed @sczyh30

Yes, the `sentinel-web-servlet` dependency could be removed.

**Comment by fangjian0423** [comment]
> If `sentinel-web-servlet` dependency is removed, `spring.cloud.sentinel.servlet` still needed?

`spring.cloud.sentinel.servlet` need remain, but this configuration item name is a little confuse to developers. Maybe `spring.cloud.sentinel.block-page` is better.

**Comment by fangjian0423** [comment]
> > If `sentinel-web-servlet` dependency is removed, `spring.cloud.sentinel.servlet` still needed?
> 
> `spring.cloud.sentinel.servlet` need remain, but this configuration item name is a little confuse to developers. Maybe `spring.cloud.sentinel.block-page` is better.

oh, sorry. I didn't know the new Class in module sentinel-spring-webmvc-adapter.

`com.alibaba.csp.sentinel.adapter.spring.webmvc.callback.UrlCleaner `
`com.alibaba.csp.sentinel.adapter.spring.webmvc.callback.BlockExceptionHandler `
`com.alibaba.csp.sentinel.adapter.spring.webmvc.callback.UrlCleaner `

These classes should be adapt to instead of the origin similar configuration items.

File: spring-cloud-alibaba-sentinel/src/main/java/com/alibaba/cloud/sentinel/SentinelProperties.java [hunk 1]
```diff
@@ -321,11 +304,11 @@ public void setClientIp(String clientIp) {
 		/**
 		 * URL pattern for sentinel filter, default is /*.
```
Comment (fangjian0423) [inline spring-cloud-alibaba-sentinel/src/main/java/com/alibaba/cloud/sentinel/SentinelProperties.java]: Add comments to describe it is for `SentinelWebInterceptor`

File: spring-cloud-alibaba-sentinel/src/main/java/com/alibaba/cloud/sentinel/SentinelWebAutoConfiguration.java [hunk 2]
```diff
@@ -61,44 +57,48 @@
...
+	@Autowired
+	private Optional<SentinelWebInterceptor> sentinelWebInterceptorOptional;
+
+	@Override
+	public void addInterceptors(InterceptorRegistry registry) {
+		if (!sentinelWebInterceptorOptional.isPresent()) {
+			return;
+		}
+		SentinelProperties.Filter filterConfig = properties.getFilter();
+		registry.addInterceptor(sentinelWebInterceptorOptional.get())
```
Comment (fangjian0423) [inline spring-cloud-alibaba-sentinel/src/main/java/com/alibaba/cloud/sentinel/SentinelWebAutoConfiguration.java]: cause NPE when `sentinelWebInterceptorOptional.get()` is null

File: spring-cloud-alibaba-sentinel/src/main/java/com/alibaba/cloud/sentinel/endpoint/SentinelEndpoint.java [hunk 3]
```diff
@@ -56,7 +55,7 @@ public SentinelEndpoint(SentinelProperties sentinelProperties) {
 			result.put("appName", AppNameUtil.getAppName());
 			result.put("logDir", LogBase.getLogBaseDir());
 			result.put("logUsePid", LogBase.isLogNameUsePid());
-			result.put("blockPage", WebServletConfig.getBlockPage());
+			result.put("blockPage", sentinelProperties.getBlockPage());
```
Comment (fangjian0423) [inline spring-cloud-alibaba-sentinel/src/main/java/com/alibaba/cloud/sentinel/endpoint/SentinelEndpoint.java]: SentinelConfig.get(BLOCK_KEY) is better

File: spring-cloud-alibaba-sentinel/src/main/resources/META-INF/additional-spring-configuration-metadata.json [hunk 4]
```diff
@@ -43,13 +43,13 @@
       "name": "spring.cloud.sentinel.filter.order",
       "type": "java.lang.Integer",
       "defaultValue": "Integer.MIN_VALUE",
-      "description": "sentinel filter chain order, will be set to FilterRegistrationBean."
+      "description": "SentinelWebInterceptor order, will be register to InterceptorRegistry."
```
Comment (fangjian0423) [inline spring-cloud-alibaba-sentinel/src/main/resources/META-INF/additional-spring-configuration-metadata.json]: It also should be added in `SentinelProperties$Filter.order`

File: spring-cloud-alibaba-sentinel/src/main/java/com/alibaba/cloud/sentinel/SentinelWebAutoConfiguration.java (hunk 2 above)
Comment (yuhuangbin) [reply spring-cloud-alibaba-sentinel/src/main/java/com/alibaba/cloud/sentinel/SentinelWebAutoConfiguration.java]: The previous code `sentinelWebInterceptorOptional.isPresent` can be avoided NPE

File: spring-cloud-alibaba-sentinel/src/main/java/com/alibaba/cloud/sentinel/SentinelProperties.java [hunk 5]
```diff
@@ -69,13 +75,7 @@
 	private Metric metric = new Metric();
 
 	/**
-	 * Web servlet configuration when the application is web, the configuration is
```
Comment (fangjian0423) [inline spring-cloud-alibaba-sentinel/src/main/java/com/alibaba/cloud/sentinel/SentinelProperties.java]: Don't remove it, refer https://github.com/spring-projects/spring-boot/blob/master/spring-boot-project/spring-boot-autoconfigure/src/main/java/org/springframework/boot/autoconfigure/web/ServerProperties.java#L142.

Comment (yuhuangbin) [reply spring-cloud-alibaba-sentinel/src/main/java/com/alibaba/cloud/sentinel/SentinelProperties.java]: If it is not deleted, do we need to support this configuration item `spring.cloud.sentinel.servlet. block-page` ?

Comment (fangjian0423) [reply spring-cloud-alibaba-sentinel/src/main/java/com/alibaba/cloud/sentinel/SentinelProperties.java]: `spring.cloud.sentinel.servlet.block-page` annotated by `DeprecatedConfigurationProperty` and `spring.cloud.sentinel.block-page` is a new item.

File: spring-cloud-alibaba-sentinel/src/main/java/com/alibaba/cloud/sentinel/SentinelWebAutoConfiguration.java [hunk 6]
```diff
@@ -16,40 +16,36 @@
...
+import org.springframework.web.servlet.config.annotation.InterceptorRegistry;
+import org.springframework.web.servlet.config.annotation.WebMvcConfigurer;
 
 /**
  * @author xiaojing
+ * @author yuhuangbin
  */
 @Configuration(proxyBeanMethods = false)
 @ConditionalOnWebApplication(type = Type.SERVLET)
-@ConditionalOnClass(CommonFilter.class)
```
Comment (echooymxq) [inline spring-cloud-alibaba-sentinel/src/main/java/com/alibaba/cloud/sentinel/SentinelWebAutoConfiguration.java]: Replace with `@ConditionalOnClass(SentinelWebInterceptor.class)`.

File: spring-cloud-alibaba-sentinel/src/main/java/com/alibaba/cloud/sentinel/SentinelProperties.java [hunk 7]
```diff
@@ -198,10 +217,14 @@ public void setColdFactor(String coldFactor) {
 		 */
 		private String blockPage;
 
+		@DeprecatedConfigurationProperty(
+				reason = "replaced to SentinelProperties#blockPage.",
+				replacement = SentinelConstants.PROPERTY_PREFIX + ".block-page")
 		public String getBlockPage() {
```
Comment (echooymxq) [inline spring-cloud-alibaba-sentinel/src/main/java/com/alibaba/cloud/sentinel/SentinelProperties.java]: Add `@Deprecated`.

File: spring-cloud-alibaba-sentinel/src/main/java/com/alibaba/cloud/sentinel/SentinelProperties.java [hunk 8]
```diff
@@ -198,10 +217,14 @@ public void setColdFactor(String coldFactor) {
 		 */
 		private String blockPage;
 
+		@DeprecatedConfigurationProperty(
+				reason = "replaced to SentinelProperties#blockPage.",
+				replacement = SentinelConstants.PROPERTY_PREFIX + ".block-page")
```
Comment (echooymxq) [inline spring-cloud-alibaba-sentinel/src/main/java/com/alibaba/cloud/sentinel/SentinelProperties.java]: Add `@Deprecated`.

**Comment by echooymxq** [comment]
I have a question，the block page will work on Servelet with `DefaultUrlBlockHandler`. Buf if we use mvc adapter, I didn't see a default handler where to use the property. Are we just retaining this field?

**Pull request merged**